File: Algorithm.Python/BracketOrderRegressionAlgorithm.py
# QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
# Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from AlgorithmImports import *

### <summary>
### Regression algorithm asserting the behavior of the bracket_order helper method (OTOCO): a market entry order
### which once filled triggers a take profit and a stop loss order, the first one to fill cancels the other
### </summary>
class BracketOrderRegressionAlgorithm(QCAlgorithm):

    def initialize(self) -> None:
        self.set_start_date(2013, 10, 7)
        self.set_end_date(2013, 10, 11)
        self.set_cash(100000)

        self._symbol = self.add_equity("SPY", Resolution.MINUTE).symbol
        self._tickets = None
        self._order_events = []

    def on_data(self, slice: Slice) -> None:
        if self._tickets is not None:
            return

        price = self.securities[self._symbol].price
        self._tickets = self.bracket_order(self._symbol, 100, take_profit_price=price * 1.005, stop_loss_price=price * 0.995, tag="Bracket")

        if len(self._tickets) != 3:
            raise RegressionTestException(f"Expected 3 order tickets, but got {len(self._tickets)}")

        entry, take_profit, stop_loss = self._tickets
        if entry.order_type != OrderType.MARKET or entry.status != OrderStatus.FILLED:
            raise RegressionTestException(f"Expected the market entry order to be filled: {entry}")
        if (take_profit.order_type != OrderType.LIMIT or take_profit.quantity != -100
            or stop_loss.order_type != OrderType.STOP_MARKET or stop_loss.quantity != -100):
            raise RegressionTestException("Unexpected take profit and stop loss orders")

        order_ids = sorted([x.order_id for x in self._tickets])
        for ticket in self._tickets:
            contingency = ticket.contingency
            if contingency is None or contingency.count != 3 or sorted(contingency.order_ids) != order_ids:
                raise RegressionTestException(f"Unexpected contingency for order {ticket.order_id}")

        entry_contingencies = entry.contingency.links
        parent = entry_contingencies[0]
        if len(entry_contingencies) != 1 or parent.type != ContingencyType.ONE_TRIGGERS_OTHER or parent.role != ContingencyRole.PARENT:
            raise RegressionTestException("Unexpected entry contingencies")

        for child in [take_profit, stop_loss]:
            contingencies = child.contingency.links
            # the entry already filled so they should of been triggered and be working
            triggered = [c for c in contingencies if c.type == ContingencyType.ONE_TRIGGERS_OTHER and c.role == ContingencyRole.CHILD
                and c.id == parent.id and c.triggered and c.triggered_time == self.utc_time]
            member = [c for c in contingencies if c.type == ContingencyType.ONE_CANCELS_OTHER and c.role is None]
            if (child.contingency.is_waiting_for_trigger or child.status != OrderStatus.SUBMITTED or len(contingencies) != 2
                or len(triggered) != 1 or len(member) != 1):
                raise RegressionTestException(f"Unexpected child order state: {child}")

    def on_order_event(self, order_event: OrderEvent) -> None:
        self._order_events.append(order_event)

    def on_end_of_algorithm(self) -> None:
        if self._tickets is None:
            raise RegressionTestException("The bracket order was never submitted")

        exits = self._tickets[1:]
        filled = [x for x in exits if x.status == OrderStatus.FILLED]
        canceled = [x for x in exits if x.status == OrderStatus.CANCELED]
        if len(filled) != 1 or len(canceled) != 1:
            raise RegressionTestException("Expected one exit to fill and the other to be canceled")

        if self.portfolio.invested:
            raise RegressionTestException("Expected the position to be closed by the bracket exit")

        # the sibling is canceled right after the fill
        fill_index = next(i for i, x in enumerate(self._order_events) if x.order_id == filled[0].order_id and x.status == OrderStatus.FILLED)
        cancel_event = self._order_events[fill_index + 1]
        if (cancel_event.order_id != canceled[0].order_id or cancel_event.status != OrderStatus.CANCELED
            or cancel_event.utc_time != self._order_events[fill_index].utc_time):
            raise RegressionTestException(f"Expected the sibling to be canceled right after the fill, but was: {cancel_event}")

        if len(self.transactions.get_open_orders()) != 0:
            raise RegressionTestException("Unexpected open orders")

File: Algorithm.CSharp/BracketOrderRegressionAlgorithm.cs
/*
 * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
 * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
*/

using System.Collections.Generic;
using System.Linq;
using QuantConnect.Data;
using QuantConnect.Interfaces;
using QuantConnect.Orders;

namespace QuantConnect.Algorithm.CSharp
{
    /// <summary>
    /// Regression algorithm asserting the behavior of the <see cref="QCAlgorithm.BracketOrder"/> helper method (OTOCO):
    /// a market entry order which once filled triggers a take profit and a stop loss order, the first one to fill cancels the other
    /// </summary>
    public class BracketOrderRegressionAlgorithm : QCAlgorithm, IRegressionAlgorithmDefinition
    {
        private Symbol _symbol;
        private List<OrderTicket> _tickets;
        private readonly List<OrderEvent> _orderEvents = new();

        /// <summary>
        /// Initialise the data and resolution required, as well as the cash and start-end dates for your algorithm. All algorithms must initialized.
        /// </summary>
        public override void Initialize()
        {
            SetStartDate(2013, 10, 07);
            SetEndDate(2013, 10, 11);
            SetCash(100000);

            _symbol = AddEquity("SPY", Resolution.Minute).Symbol;
        }

        /// <summary>
        /// OnData event is the primary entry point for your algorithm. Each new data point will be pumped in here.
        /// </summary>
        /// <param name="slice">Slice object keyed by symbol containing the stock data</param>
        public override void OnData(Slice slice)
        {
            if (_tickets != null)
            {
                return;
            }

            var price = Securities[_symbol].Price;
            _tickets = BracketOrder(_symbol, 100, takeProfitPrice: price * 1.005m, stopLossPrice: price * 0.995m, tag: "Bracket");

            if (_tickets.Count != 3)
            {
                throw new RegressionTestException($"Expected 3 order tickets, but got {_tickets.Count}");
            }

            var entry = _tickets[0];
            var takeProfit = _tickets[1];
            var stopLoss = _tickets[2];
            if (entry.OrderType != OrderType.Market || entry.Status != OrderStatus.Filled)
            {
                throw new RegressionTestException($"Expected the market entry order to be filled: {entry}");
            }
            if (takeProfit.OrderType != OrderType.Limit || takeProfit.Quantity != -100 || stopLoss.OrderType != OrderType.StopMarket || stopLoss.Quantity != -100)
            {
                throw new RegressionTestException("Unexpected take profit and stop loss orders");
            }

            foreach (var ticket in _tickets)
            {
                if (ticket.Contingency == null || ticket.Contingency.Count != 3
                    || !ticket.Contingency.OrderIds.SetEquals(_tickets.Select(x => x.OrderId)))
                {
                    throw new RegressionTestException($"Unexpected contingency for order {ticket.OrderId}");
                }
            }

            var parent = entry.Contingency.Links.Single();
            if (parent.Type != ContingencyType.OneTriggersOther || parent.Role != ContingencyRole.Parent)
            {
                throw new RegressionTestException($"Unexpected entry contingencies: {string.Join(",", entry.Contingency.Links)}");
            }

            foreach (var child in new[] { takeProfit, stopLoss })
            {
                // the entry already filled so they should of been triggered and be working
                if (child.Contingency.IsWaitingForTrigger || child.Status != OrderStatus.Submitted || child.Contingency.Links.Count != 2
                    || !child.Contingency.Links.Any(link => link.Type == ContingencyType.OneTriggersOther && link.Role == ContingencyRole.Child && link.Id == parent.Id
                        && link.Triggered && link.TriggeredTime == UtcTime)
                    || !child.Contingency.Links.Any(link => link.Type == ContingencyType.OneCancelsOther && link.Role == null))
                {
                    throw new RegressionTestException($"Unexpected child order state: {child}. Contingencies: {string.Join(",", child.Contingency.Links)}");
                }
            }
        }

        /// <summary>
        /// Order event handler
        /// </summary>
        public override void OnOrderEvent(OrderEvent orderEvent)
        {
            _orderEvents.Add(orderEvent);
        }

        /// <summary>
        /// End of algorithm run event handler
        /// </summary>
        public override void OnEndOfAlgorithm()
        {
            if (_tickets == null)
            {
                throw new RegressionTestException("The bracket order was never submitted");
            }

            var exits = _tickets.Skip(1).ToList();
            var filled = exits.Where(x => x.Status == OrderStatus.Filled).ToList();
            var canceled = exits.Where(x => x.Status == OrderStatus.Canceled).ToList();
            if (filled.Count != 1 || canceled.Count != 1)
            {
                throw new RegressionTestException($"Expected one exit to fill and the other to be canceled: {string.Join(" | ", exits)}");
            }

            if (Portfolio.Invested)
            {
                throw new RegressionTestException("Expected the position to be closed by the bracket exit");
            }

            // the sibling is canceled right after the fill
            var fillIndex = _orderEvents.FindIndex(x => x.OrderId == filled[0].OrderId && x.Status == OrderStatus.Filled);
            var cancelEvent = _orderEvents[fillIndex + 1];
            if (cancelEvent.OrderId != canceled[0].OrderId || cancelEvent.Status != OrderStatus.Canceled || cancelEvent.UtcTime != _orderEvents[fillIndex].UtcTime)
            {
                throw new RegressionTestException($"Expected the sibling to be canceled right after the fill, but was: {cancelEvent}");
            }

            if (Transactions.GetOpenOrders().Count != 0)
            {
                throw new RegressionTestException("Unexpected open orders");
            }
        }

        /// <summary>
        /// This is used by the regression test system to indicate if the open source Lean repository has the required data to run this algorithm.
        /// </summary>
        public bool CanRunLocally { get; } = true;

        /// <summary>
        /// This is used by the regression test system to indicate which languages this algorithm is written in.
        /// </summary>
        public List<Language> Languages { get; } = new() { Language.CSharp, Language.Python };

        /// <summary>
        /// Data Points count of all timeslices of algorithm
        /// </summary>
        public long DataPoints => 3943;

        /// <summary>
        /// Data Points count of the algorithm history
        /// </summary>
        public int AlgorithmHistoryDataPoints => 0;

        /// <summary>
        /// Final status of the algorithm
        /// </summary>
        public AlgorithmStatus AlgorithmStatus => AlgorithmStatus.Completed;

        /// <summary>
        /// This is used by the regression test system to indicate what the expected statistics are from running the algorithm
        /// </summary>
        public Dictionary<string, string> ExpectedStatistics => new Dictionary<string, string>
        {
            {"Total Orders", "3"},
            {"Average Win", "0.07%"},
            {"Average Loss", "0%"},
            {"Compounding Annual Return", "5.586%"},
            {"Drawdown", "0.000%"},
            {"Expectancy", "0"},
            {"Start Equity", "100000"},
            {"End Equity", "100069.52"},
            {"Net Profit", "0.070%"},
            {"Sharpe Ratio", "0"},
            {"Sortino Ratio", "0"},
            {"Probabilistic Sharpe Ratio", "0%"},
            {"Loss Rate", "0%"},
            {"Win Rate", "100%"},
            {"Profit-Loss Ratio", "0"},
            {"Alpha", "0"},
            {"Beta", "0"},
            {"Annual Standard Deviation", "0"},
            {"Annual Variance", "0"},
            {"Information Ratio", "-8.91"},
            {"Tracking Error", "0.223"},
            {"Treynor Ratio", "0"},
            {"Total Fees", "$2.00"},
            {"Estimated Strategy Capacity", "$31000000.00"},
            {"Lowest Capacity Asset", "SPY R735QTJ8XC9X"},
            {"Portfolio Turnover", "5.80%"},
            {"Drawdown Recovery", "0"},
            {"OrderListHash", "4ab291d7e7df4d2d944da910653113b6"}
        };
    }
}
